File: wallet-service/src/main/kotlin/commands/domain/models/base/AggregateRoot.kt
package com.taager.wallet.commands.domain.models.base
abstract class AggregateRoot<out T : ValueObject>(id: T) : Entity<T>(id)
{
    protected val occurredEvents: MutableList<DomainEvent> = mutableListOf()
    fun occurredEvents(): List<DomainEvent>
    {
        val events = this.occurredEvents.toMutableList()
        this.occurredEvents.clear()
        return events
    }
    protected fun raiseEvent(event: DomainEvent)
    {
        occurredEvents.add(event)
    }
}

File: Merchant-service/src/shared-kernel/domain/base/AggregateRoot.ts
export interface DomainEvent {}
export interface ValueObject {}
export interface DomainEntity {}
export abstract class AggregateRoot {
  private occurredEvents: DomainEvent[] = [];
  raiseEvent(domainEvent: DomainEvent) {
    this.occurredEvents.push(domainEvent);
  }
  getOccurredEvents(): DomainEvent[] {
    return [...this.occurredEvents];
  }

  public clone<T>(): T {
    return <T>JSON.parse(JSON.stringify(this));
  }
}


